### ISSUE TITLE
Render rooftop solar above buildings

### ISSUE THREAD

**Issue by arrival-spring** [gravitystorm/openstreetmap-carto#4364]
### Expected behavior
`power=generator` with generator:source=solar` (and `location=roof`) should render above buildings.
### Actual behavior
It renders below buildings. In an area near me, there are some buildings mapped and also some rooftop solar mapped as areas. At the moment, the rooftop solar can only be seen when the building that it is on has not been mapped.

**Comment by imagico** [comment]
About 350k combinations of power=generator + location=roof.

In principle this would be possible but i have doubts if this would be a good idea considering it would be something of low practical relevance for typical map uses at the highest zoom levels (which is the only case where this would be feasible).  But final assessment would depend on the specific design suggested.

**Comment by HolgerJeromin** [comment]
Not an easy prioritization problem: #3435

**Comment by imagico** [comment]
No, power=generator is rendered with a polygons fill and therefore is a non-blocking element (in contrast to addresses/POI symbols).

**Comment by Hufkratzer** [comment]
> About 350k combinations of power=generator + location=roof.
> 
> In principle this would be possible but i have doubts if this would be a good idea considering it would be something of low practical relevance for typical map uses at the highest zoom levels (which is the only case where this would be feasible). But final assessment would depend on the specific design suggested.

Probably related: https://github.com/gravitystorm/openstreetmap-carto/issues/4443 (swimming pools)

**Comment by pnorman** [comment]
> `power=generator` with `generator:source=solar` (and `location=roof`) should render above buildings.

In practical terms, all power=generator must be below or above buildings in the order. What would this change do to other power=generator?

**Comment by russss** [comment]
A presumably related issue is that floating solar generators [should be rendered on top of water areas](https://community.openstreetmap.org/t/solar-panels-over-water/124831).

> In practical terms, all power=generator must be below or above buildings in the order. What would this change do to other power=generator?

I think it would be fine to render generators on top of buildings. I can't think of a valid tagging scenario where a building would be inside a generator. I'm fairly confident that in all cases, generators should either be at a smaller scale than buildings, or they should be tagged on the same way as the building. (This is in contrast to `power=plant`s, which often do contain buildings.)

**Comment by dch0ph** [comment]
> I think it would be fine to render generators on top of buildings.

Paul Norman's point is about the order in which layers are drawn, rather than whether the generators are physically above / below buildings. 

I think the current status quo is fine, with the buildings layer coming after / above power generators etc. 

I think adding a special case for rooftop solar generation would add a lot of visual noise on a general map.

**Comment by russss** [comment]
> Paul Norman's point is about the order in which layers are drawn, rather than whether the generators are physically above / below buildings.

So was my reply. I don't think there's any valid tagging scenario where a `power=generator` area would obscure a building area, so from that perspective I don't think there's any issue rendering all generators on top of buildings - it wouldn't need a special case.

From a cartographic perspective I'm not sure how well it'd work with Carto without seeing it.

**Comment by dch0ph** [comment]
Sorry for the crossed wires.

But the point remains that rendering `power=generator` above buildings would require us to remove it from the "landuse" layer and create a new "special landuse" layer above buildings. This would be hard to justify.

You could carve out an exception for `location=rooftop`, which could handle other cases such as rooftop swimming pools, but as mentioned above, I don't think there will be much enthusiasm for including solar panels in a "rooftop layer".
